src/test/regress/output/temp_tablespaces.source

```diff
@@ -227,3 +270,300 @@ drop tablespace mytempsp1;
+\! gpstop -u
+-- end_ignore
+SELECT * FROM gp_tablespace_file_report(1);
+ temp_table_in | temp_files_in 
+---------------+---------------
+ default       | temp_filesN
+ default       | temp_filesN
+ default       | temp_filesN
+(3 rows)
+
+-- Files should be in temp_files0, table should be in temp_tables.
+-- start_ignore
+\! gpconfig -c temp_tablespaces -v 'temp_tables'
+\! gpconfig -c temp_spill_files_tablespaces -v 'temp_files0'
+\! gpstop -u
+-- end_ignore
+SELECT * FROM gp_tablespace_file_report(1);
+ temp_table_in | temp_files_in 
+---------------+---------------
+ temp_tables   | temp_filesN
+ temp_tables   | temp_filesN
+ temp_tables   | temp_filesN
+(3 rows)
+
+-- Files should be in temp_filesN, where N = gp_session_id % 4, table should be
+-- in the default tablespace.
+-- start_ignore
+\! gpconfig -r 'temp_tablespaces'
+\! gpconfig -c temp_spill_files_tablespaces -v 'temp_files0','temp_files1','temp_files2','temp_files3'
+\! gpstop -u
+-- end_ignore
+SELECT * FROM gp_tablespace_file_report(4);
+ temp_table_in | temp_files_in 
+---------------+---------------
+ default       | temp_filesN
+ default       | temp_filesN
+ default       | temp_filesN
+(3 rows)
+
+-- Files should be in the default tablespace, table should be in temp_tables.
+-- start_ignore
+\! gpconfig -c temp_tablespaces -v 'temp_tables'
+\! gpconfig -c temp_spill_files_tablespaces -v '""'
+\! gpstop -u
+-- end_ignore
+SELECT * FROM gp_tablespace_file_report(1);
+ temp_table_in | temp_files_in 
+---------------+---------------
+ temp_tables   | default
+ temp_tables   | default
+ temp_tables   | default
+(3 rows)
+
+-- Files should be in the default tablespace, table should be in temp_tables.
+-- start_ignore
+\! gpconfig -c temp_tablespaces -v 'temp_tables'
+\! gpconfig -c temp_spill_files_tablespaces -v 'pg_default'
+\! gpstop -u
+-- end_ignore
+SELECT * FROM gp_tablespace_file_report(1);
+ temp_table_in | temp_files_in 
+---------------+---------------
+ temp_tables   | default
+ temp_tables   | default
+ temp_tables   | default
+(3 rows)
+
+-- Reset the global config before testing session-only values.
+-- start_ignore
+\! gpconfig -r temp_tablespaces
+\! gpconfig -r temp_spill_files_tablespaces
+\! gpstop -u
+-- end_ignore
+--
+-- Session-only GUC value tests:
+--
+-- Both files and the table should be in the default tablespace.
+RESET temp_tablespaces;
+RESET temp_spill_files_tablespaces;
+SELECT * FROM gp_tablespace_file_report(1);
+ temp_table_in | temp_files_in 
+---------------+---------------
+ default       | default
+ default       | default
+ default       | default
+(3 rows)
+
+-- Both files and the table should be in temp_tables.
+SET temp_tablespaces = 'temp_tables';
+RESET temp_spill_files_tablespaces;
+SELECT * FROM gp_tablespace_file_report(1);
+ temp_table_in | temp_files_in 
+---------------+---------------
+ temp_tables   | temp_tables
+ temp_tables   | temp_tables
+ temp_tables   | temp_tables
+(3 rows)
```

Comment:
The tests failed here.

Reply:
This seems to be another GUC-related issue. At some point I've managed to achieve
```
\! gpconfig -c temp_tablespaces -v 'temp_tables'
\! gpstop -u
SHOW temp_tablespaces;
 temp_tablespaces 
------------------
 temp_tables
(1 row)

\! gpconfig -r temp_tablespaces
\! gpstop -u
SHOW temp_tablespaces;
 temp_tablespaces 
------------------
 temp_tables
(1 row)
```
while testing locally, although I can't replicate this reliably yet. I'll research this more.

Note: leaving the last `\c` makes the test pass.

Reply:
Yup, GUC sync issue. Reproduction scenario (both on GPDB 6 and GPDB 7):
```
SET temp_tablespaces = 'pg_default';
SHOW temp_tablespaces;
\! gpconfig -r temp_tablespaces
\! gpstop -u
SHOW temp_tablespaces;
```
outputs `pg_default` on both SHOWs.

Reply:
This is indeed a GUC sync issue. After GUC reset, coordinator sends `SET temp_spill_files_tablespaces TO ''` to segments, except in case of this specific GUC `RESET` is not equivalent to setting the default value (empty string) explicitly. In general this behavior seems very confusing.

Reply:
Fixed this issue for GPDB 6, cherry picked the commit here.